src/test/java/org/opensearch/security/auditlog/integration/TestAuditlogImpl.java

```diff
@@ -25,23 +28,60 @@
 
 public class TestAuditlogImpl extends AuditLogSink {
 
+    /** Use the results of `doThenWaitForMessages(...)` instead */
+    @Deprecated
     public static List<AuditMessage> messages = new ArrayList<AuditMessage>(100);
+    /** Check messages indvidually instead of searching this string */
+    @Deprecated
     public static StringBuffer sb = new StringBuffer();
+    private static final AtomicReference<CountDownLatch> countDownRef = new AtomicReference<>();
+    private static final AtomicReference<List<AuditMessage>> messagesRef = new AtomicReference<>();
 
     public TestAuditlogImpl(String name, Settings settings, String settingsPrefix, AuditLogSink fallbackSink) {
         super(name, settings, null, fallbackSink);
     }
 
-
-    public synchronized boolean  doStore(AuditMessage msg) {
+    public boolean doStore(AuditMessage msg) {
+        if (messagesRef.get() == null || countDownRef.get() == null) {
+            throw new RuntimeException("No message latch is waiting");
+        }
         sb.append(msg.toPrettyString()+System.lineSeparator());
-        messages.add(msg);
+        messagesRef.get().add(msg);
+        countDownRef.get().countDown();
         return true;
     }
 
+    /** Unneeded after switching to `doThenWaitForMessages(...)` as data is automatically flushed */
+    @Deprecated
     public static synchronized void clear() {
-        sb.setLength(0);
-        messages.clear();
+        doThenWaitForMessages(() -> {}, 0);
+    }
+
+    /**
+     * Perform an action and then wait until the expected number of messages have been found.
+     */
+    public static List<AuditMessage> doThenWaitForMessages(final Runnable action, final int expectedCount) {
+        final CountDownLatch latch = new CountDownLatch(expectedCount);
+        final List<AuditMessage> messages = new ArrayList<>();
+        countDownRef.set(latch);
+        messagesRef.set(messages);
+
+        TestAuditlogImpl.sb = new StringBuffer();
+        TestAuditlogImpl.messages = messages; 
+        
+        try {
+            action.run();
+            final boolean foundAll = latch.await(1, TimeUnit.SECONDS);
```

Comment:
nit: magic number, perhaps extract to constant. Not used anywhere else so not that big of a deal, up to you.